feat: add module build script

manzt opened the pull request:
Bundles `src/index.js` with esbuild. The bundle generated is intended to be consumed by other packages that require `vitessce` as a dependency. The target therefore does not bundle dependencies, and instead only:

- compiles sass -> .css
- inlines worker module
- imports `.svg` as `ReactComponent`

manzt commented on src/css/_higlass.scss:
```diff
@@ -2,7 +2,7 @@
   // https://sass-lang.com/documentation/at-rules/import#nesting
   // https://sass-lang.com/documentation/at-rules/import#importing-css
-  @import "~higlass/dist/hglib";
+  @import "../../node_modules/higlass/dist/hglib";
```
`esbuild-plugin-sass` couldn't seem to resolve the ~ alias. I changed it to a relative import and the webpack dev-server seemed to still work (it also made the esbuild plugin happy).

manzt commented:
Although some things aren't working, I built #1045 using this script and the simple dynamic imports are supported

![image](https://user-images.githubusercontent.com/24403730/133849432-70e8875d-1595-4fd9-ab18-a7c481ea12f4.png)

ilan-gold commented:
@keller-mark is this used though?  I though they were compiled directly via `glsl` as strings.  Possible I ma misunderstanding this.

ilan-gold commented:
Relatedly @manzt I did look at this but ran into the issues you ran in to above.  Do you have any ideas of what you might want to do to resolve them?  I am trying out upgrading `webpack` but it seems like there are easier solutions, at least for the transpiling issue.

manzt commented:
@ilan-gold I have some changes and will push in a bit.

manzt commented:
@ilan-gold. I added an esbuild plugin to transform _some_ files with babel. This is inefficient compared to esbuild's native transforms, so I added a filter to restrict this plugin to only files with "shader" in the filename. In addition, I had to convert the build script to a module ('.mjs') since `esbuild-plugin-babel` is ESM only (can't load via commonjs `require()`).

As a side node, if we changed shader imports from babel-transpiled javascript modules to static assets (e.g. `import fs from './fs.glsl`) then we wouldn't need to have the babel plugin at all. However, the goal of this PR is avoid touching the source code or any other webpack config. This is something we can look into in the future but doesn't need to be address at the moment IMO.

Also, Babel seems to require that a `NODE_ENV` or `BABEL_ENV` flag are set in the environment when transpiling. I am not sure what to set this to when building our target, but for now you will need to run:


```
NODE_ENV=development node script/build-esm.mjs
```

ilan-gold commented:
@manzt Trying it out now but I got an error on this latest commit:
```
esbuild.build({
        ^

TypeError: esbuild.build is not a function
    at file:///Users/ilan_gold/gehlenborg/vitessce/scripts/build-esm.mjs:107:9
    at ModuleJob.run (internal/modules/esm/module_job.js:140:23)
    at async Loader.import (internal/modules/esm/loader.js:165:24)
    at async Object.loadESM (internal/process/esm_loader.js:68:5)
```
which seeemed to be fixed by using 
```javascript
import esbuild from "esbuild";
```
instead of
```javascript
import * as esbuild from "esbuild";
```

manzt commented:
It _should_ work. What version of node are you using?

ilan-gold commented:
@manzt I am on 14.7.  Besides this, I was able to load the `higlass` component in the portal without issue.

keller-mark commented:
>  In addition, I had to convert the build script to a module ('.mjs') since esbuild-plugin-babel is ESM only

Is the `build-esm.js` script no longer needed?

keller-mark commented:
I experienced the same error as Ilan, I am on Node v14.0.0

ilan-gold commented:
@manzt One thing I notice here is that this seems to push the `txml` problem further up the stack, which I think is not desirable since people don't even know where `txml` is coming from.  It seems like merging the recent `geotiff` shift away from it: https://github.com/geotiffjs/geotiff.js/pull/245 and then upgrading Viv would be the long term solution (before any release of this branch).  Seems like not a blocker, but if you have a quick fix for this here, that would of course be welcome.

As for the node v10 support, I wonder how many people are still using it.  I'd be fine dropping it.
